src/mount.rs

```diff
@@ -68,42 +66,47 @@ mod ffi {
     }
 }
 
-pub fn mount(
-        source: Option<&Path>,
-        target: &Path,
-        fstype: Option<&str>,
+// XXX: Should `data` be a `NixPath` here?
+pub fn mount<P1: NixPath, P2: NixPath, P3: NixPath, P4: NixPath>(
+        source: Option<P1>,
+        target: P2,
+        fstype: Option<P3>,
         flags: MsFlags,
-        data: Option<&str>) -> SysResult<()> {
-
-    let source = source.map(|s| s.to_c_str());
-    let target = target.to_c_str();
-    let fstype = fstype.map(|s| s.to_c_str());
-    let data = data.map(|s| s.to_c_str());
-
-    let res = unsafe {
-        ffi::mount(
-            source.as_ref().map(|s| s.as_ptr()).unwrap_or(ptr::null()),
-            target.as_ptr(),
-            fstype.as_ref().map(|s| s.as_ptr()).unwrap_or(ptr::null()),
-            flags.bits,
-            data.map(|s| s.as_ptr() as *const c_void).unwrap_or(ptr::null()))
-    };
-
-    from_ffi(res)
+        data: Option<P4>) -> NixResult<()> {
+    use libc;
+
+    let res = try!(try!(try!(try!(
```

Comment:
ewww yeah... there probably is a way to clean it up, but I wouldn't be too worried about it ATM. This is ok for now